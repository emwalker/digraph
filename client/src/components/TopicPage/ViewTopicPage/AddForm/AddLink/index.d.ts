import React from 'react';
import { AddLink_viewer$key } from '__generated__/AddLink_viewer.graphql';
import { AddLink_parentTopic$key } from '__generated__/AddLink_parentTopic.graphql';
type Props = {
    disabled?: boolean;
    parentTopic: AddLink_parentTopic$key;
    viewer: AddLink_viewer$key;
};
export default function AddLink(props: Props): React.JSX.Element;
export {};
