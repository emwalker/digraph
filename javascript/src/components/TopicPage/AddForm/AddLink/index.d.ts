import { RelayProp } from 'react-relay';
import { AddLink_viewer$data as ViewerType } from '__generated__/AddLink_viewer.graphql';
import { AddLink_topic$data as TopicType } from '__generated__/AddLink_topic.graphql';
declare type Props = {
    disabled?: boolean;
    relay: RelayProp;
    topic: TopicType;
    viewer: ViewerType;
};
declare const _default: import("react-relay").Container<Pick<Omit<Props, "relay">, "topic" | "viewer"> & {
    disabled?: boolean | undefined;
} & {}>;
export default _default;
