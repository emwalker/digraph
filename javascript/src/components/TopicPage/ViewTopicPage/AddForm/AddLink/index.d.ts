import { AddLink_viewer$key } from '__generated__/AddLink_viewer.graphql';
import { AddLink_parentTopic$key } from '__generated__/AddLink_parentTopic.graphql';
declare type Props = {
    disabled?: boolean;
    parentTopic: AddLink_parentTopic$key;
    viewer: AddLink_viewer$key;
};
export default function AddLink(props: Props): JSX.Element;
export {};
