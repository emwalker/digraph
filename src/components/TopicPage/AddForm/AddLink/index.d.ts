import { RelayProp } from 'react-relay';
import { AddLink_viewer as ViewerType } from '__generated__/AddLink_viewer.graphql';
import { AddLink_topic as TopicType } from '__generated__/AddLink_topic.graphql';
declare type Props = {
    disabled?: boolean;
    relay: RelayProp;
    topic: TopicType;
    viewer: ViewerType;
};
declare const _default: import("react-relay").Container<Pick<Omit<Props, "relay">, "topic" | "viewer"> & {
    disabled?: boolean | undefined;
} & {}>;
export default _default;
