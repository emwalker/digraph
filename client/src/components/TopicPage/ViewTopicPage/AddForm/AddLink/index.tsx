import React, {
  FormEvent, useState, useCallback, KeyboardEvent,
} from 'react'
import { graphql, useFragment } from 'react-relay'

import { makeUpsertLinkCallback } from 'mutations/upsertLinkMutation'
import {
  AddLink_viewer$key,
} from '__generated__/AddLink_viewer.graphql'
import {
  AddLink_parentTopic$key,
} from '__generated__/AddLink_parentTopic.graphql'

const tooltip = 'Add a link to this topic.\n'
  + 'Press "Return" to submit the new link.'

type Props = {
  disabled?: boolean,
  parentTopic: AddLink_parentTopic$key,
  viewer: AddLink_viewer$key,
}

const viewerFragment = graphql`
  fragment AddLink_viewer on User {
    selectedRepoId
  }
`

const topicFragment = graphql`
  fragment AddLink_parentTopic on Topic {
    id
  }
`

export default function AddLink(props: Props) {
  const viewer = useFragment(viewerFragment, props.viewer)
  const parentTopic = useFragment(topicFragment, props.parentTopic)
  const [url, setUrl] = useState('')
  const selectedRepoId = viewer.selectedRepoId
  const upsertLink = makeUpsertLinkCallback({
    url, setUrl, topicId: parentTopic.id, selectedRepoId,
  })

  const onKeyPress = useCallback((event: KeyboardEvent<HTMLInputElement>) => {
    if (event.key !== 'Enter') return
    upsertLink()
  }, [upsertLink])

  const onChange = useCallback((event: FormEvent<HTMLInputElement>) => {
    setUrl(event.currentTarget.value)
  }, [setUrl])

  return (
    <dl className="form-group">
      <dt>
        <span
          className="tooltipped tooltipped-ne"
          aria-label={tooltip}
        >
          <label htmlFor="create-link-url">Add link</label>
        </span>
      </dt>
      <dd>
        <input
          className="form-control input-sm"
          disabled={props.disabled}
          data-testid="link-url-input"
          id="create-link-url"
          onChange={onChange}
          onKeyPress={onKeyPress}
          placeholder="Url"
          type="url"
          value={url}
        />
      </dd>
    </dl>
  )
}
