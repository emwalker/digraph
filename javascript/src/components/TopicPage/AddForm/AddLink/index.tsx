import React, { Component, KeyboardEvent, FormEvent } from 'react'
import { graphql, createFragmentContainer, RelayProp, DeclarativeMutationConfig } from 'react-relay'

import upsertLinkMutation, { Input } from 'mutations/upsertLinkMutation'
import { AddLink_viewer as ViewerType } from '__generated__/AddLink_viewer.graphql'
import { AddLink_topic as TopicType } from '__generated__/AddLink_topic.graphql'

type RepositoryType = ViewerType['selectedRepository']

const tooltip = 'Add a link to this topic.\n'
  + 'Press "Return" to submit the new link.'

type Props = {
  disabled?: boolean,
  relay: RelayProp,
  topic: TopicType,
  viewer: ViewerType,
}

type State = {
  url: string,
}

class AddLink extends Component<Props, State> {
  static defaultProps = {
    disabled: false,
  }

  constructor(props: Props) {
    super(props)
    this.state = {
      url: '',
    }
  }

  onKeyPress = (event: KeyboardEvent<HTMLInputElement>) => {
    if (event.key === 'Enter') this.createLink()
  }

  get selectedRepo(): RepositoryType {
    return this.props.viewer.selectedRepository
  }

  get orgLogin() {
    const repo = this.selectedRepo
    if (!repo) return null
    return repo.organization.login
  }

  get relayConfigs(): DeclarativeMutationConfig[] {
    return [{
      type: 'RANGE_ADD',
      parentID: this.props.topic.id,
      connectionInfo: [{
        key: 'Topic_links',
        rangeBehavior: 'prepend',
      }],
      edgeName: 'linkEdge',
    }]
  }

  updateUrl = (event: FormEvent<HTMLInputElement>) => {
    this.setState({ url: event.currentTarget.value })
  }

  createLink() {
    const repo = this.selectedRepo
    const repoName = repo ? repo.name : null
    const { orgLogin } = this

    if (!repoName) return
    if (!orgLogin) return

    const input: Input = {
      addParentTopicIds: [this.props.topic.id],
      organizationLogin: orgLogin,
      repositoryName: repoName,
      url: this.state.url,
    }

    upsertLinkMutation(
      this.props.relay.environment,
      input,
      {
        configs: this.relayConfigs,
      },
    )
    this.setState({ url: '' })
  }

  render = () => (
    <dl className="form-group">
      <dt>
        <span
          className="tooltipped tooltipped-ne"
          aria-label={tooltip}
        >
          <label htmlFor="create-link-url">Add link</label>
        </span>
      </dt>
      <dd>
        <input
          className="form-control test-link-url input-sm"
          disabled={this.props.disabled}
          id="create-link-url"
          onChange={this.updateUrl}
          onKeyPress={this.onKeyPress}
          placeholder="Url"
          type="url"
          value={this.state.url}
        />
      </dd>
    </dl>
  )
}

export default createFragmentContainer(AddLink, {
  viewer: graphql`
    fragment AddLink_viewer on User {
      selectedRepository {
        id
        name

        organization {
          login
        }
      }
    }
  `,
  topic: graphql`
    fragment AddLink_topic on Topic {
      id
    }
  `,
})
