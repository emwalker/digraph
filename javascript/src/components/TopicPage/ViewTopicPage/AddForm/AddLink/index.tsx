import React, {
  FormEvent, useState, useCallback, KeyboardEvent,
} from 'react'
import { graphql, useFragment } from 'react-relay'

import { makeUpsertLinkCallback } from 'mutations/upsertLinkMutation'
import {
  AddLink_viewer$key,
} from '__generated__/AddLink_viewer.graphql'
import {
  AddLink_topic$key,
} from '__generated__/AddLink_topic.graphql'

const tooltip = 'Add a link to this topic.\n'
  + 'Press "Return" to submit the new link.'

type Props = {
  disabled?: boolean,
  topic: AddLink_topic$key,
  viewer: AddLink_viewer$key,
}

const viewerFragment = graphql`
  fragment AddLink_viewer on User {
    selectedRepoId
  }
`

const topicFragment = graphql`
  fragment AddLink_topic on Topic {
    id
  }
`

export default function AddLink(props: Props) {
  const viewer = useFragment(viewerFragment, props.viewer)
  const topic = useFragment(topicFragment, props.topic)
  const [url, setUrl] = useState('')
  const selectedRepoId = viewer.selectedRepoId
  const upsertLink = makeUpsertLinkCallback({ url, setUrl, topicId: topic.id, selectedRepoId })

  const onKeyPress = useCallback((event: KeyboardEvent<HTMLInputElement>) => {
    if (event.key !== 'Enter') return
    upsertLink()
  }, [upsertLink])

  const onChange = useCallback((event: FormEvent<HTMLInputElement>) => {
    setUrl(event.currentTarget.value)
  }, [setUrl])

  return (
    <dl className="form-group">
      <dt>
        <span
          className="tooltipped tooltipped-ne"
          aria-label={tooltip}
        >
          <label htmlFor="create-link-url">Add link</label>
        </span>
      </dt>
      <dd>
        <input
          className="form-control input-sm"
          disabled={props.disabled}
          data-testid="link-url-input"
          id="create-link-url"
          onChange={onChange}
          onKeyPress={onKeyPress}
          placeholder="Url"
          type="url"
          value={url}
        />
      </dd>
    </dl>
  )
}
