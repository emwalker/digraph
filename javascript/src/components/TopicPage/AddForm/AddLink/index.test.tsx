import React from 'react'
import { shallow } from 'enzyme'

import AddLink from './index'

jest.mock('react-relay', () => ({ createFragmentContainer: (component: any) => component }))

const props = {
  disabled: false,
  relay: { environment: {} },
  topic: {},
  viewer: {},
}

describe('<AddLink />', () => {
  // @ts-expect-error
  const wrapper = shallow(<AddLink {...props} />)

  it('renders', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('has a tooltip', () => {
    expect(wrapper.find('.tooltipped').exists()).toBeTruthy()
  })
})
