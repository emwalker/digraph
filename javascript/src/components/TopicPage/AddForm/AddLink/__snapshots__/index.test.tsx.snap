// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddLink /> renders 1`] = `
<dl
  className="form-group"
>
  <dt>
    <span
      aria-label="Add a link to this topic.
Press \\"Return\\" to submit the new link."
      className="tooltipped tooltipped-ne"
    >
      <label
        htmlFor="create-link-url"
      >
        Add link
      </label>
    </span>
  </dt>
  <dd>
    <input
      className="form-control test-link-url input-sm"
      disabled={false}
      id="create-link-url"
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder="Url"
      type="url"
      value=""
    />
  </dd>
</dl>
`;
